File: simulation/GetLaser.m
function [dis] = GetLaser(robot_X,robot_Y,robot_t,Eff_robot,K,MODE)
% Coverage MODE
% 1:non-overlapping(break), 2: over-lapping 3: non-overlapping

N=200;
map=zeros(N,N);

robot_N=size(robot_X,2);
grid_resolution=10; % 2 for IGCL, 5 for URL
sensor.r=100;%400/grid_resolution;%80;%80;
degree=60;%120;%60;
sensor.phi=degree;%;degree+1;
sensor.scan=5;%degree*4;
figure(2);
for Iteration=1:1
    %close all;

    % obstacle setting
    map=Map_creator(N);
    % plot lines & compute coverage
    for i=1:robot_N
    position(i,1)=robot_X(i);   
    position(i,2)=robot_Y(i);
    position(i,3)=robot_t(i);   
    robot.x=position(i,1);
    robot.y=position(i,2);
    robot.theta=position(i,3);
        if Eff_robot(i)==1
        [map,dis]= PlotLine(robot,sensor,map,MODE);
        end
    end
    figure(2);
    [Ratio,Coverage_rate,NONOL_Coverage_rate,OL_Coverage_rate]=PlotMap(map);hold on;
    Dir_r=4;
    for i=1:robot_N
        if Eff_robot(i)==1
        plot(position(i,1),position(i,2),'go');hold on;     
        line_x=[position(i,1) position(i,1)+Dir_r*cos(position(i,3))];
        line_y=[position(i,2) position(i,2)+Dir_r*sin(position(i,3))];
        line(line_x,line_y,'Color',[0 1 0]);   
        text(position(i,1)+5,position(i,2),num2str(i));
        else
        plot(position(i,1),position(i,2),'rx');hold on;    
        text(position(i,1)+5,position(i,2),num2str(i));        
        end
    end
    for i=1:robot_N
    X_data(1,1+(i-1)*3:3+(i-1)*3)=[position(i,1) position(i,2) position(i,3)];
    end
    X(Iteration,:)=X_data;
    Coverage(Iteration,1)=Coverage_rate;
    hold off;
    Eff_N=sum(Eff_robot);hold on; 
    %title(['Iteration=',num2str(K),', Coverage=',num2str(Coverage_rate),', Ratio=',num2str(Ratio),', Overlap=',num2str(OL_Coverage_rate), ', Eff_N=',num2str(Eff_N)])
    %fprintf('coverage   =%f\nRatio=%f\noverlap    =%f\n',Coverage_rate,Ratio,OL_Coverage_rate);
    %title([ 'Coverage=',num2str(Coverage_rate)]);
end

File: real/src/Getlaser.py
import numpy as np
import matplotlib.pyplot as plt

def GetLaser(robot_X, robot_Y, robot_t, Eff_robot, K, MODE):
    """
    Simulates laser sensor readings for a robot
    Args:
        robot_X: x positions (array-like)
        robot_Y: y positions (array-like)
        robot_t: orientations (array-like)
        Eff_robot: effectiveness flags (array-like)
        K: iteration number (unused in this simplified version)
        MODE: coverage mode (1: non-overlapping break, 2: overlapping, 3: non-overlapping)
    Returns:
        dis: laser distance readings (numpy array)
    """
    N = 200
    map = np.zeros((N, N))
    robot_N = len(robot_X)
    grid_resolution = 10
    sensor = {'r': 100, 'phi': 60, 'scan': 5}  # sensor parameters
    
    # Initialize position array
    position = np.column_stack((robot_X, robot_Y, robot_t))
    
    # Dummy laser readings - in a real implementation, this would come from actual sensor simulation
    # Here we return fixed values that would represent distances in 5 directions
    dis = np.array([100.0] * 5)  # 5 laser readings
    
    # Visualization (optional)
    if False:  # Set to True to enable plotting
        plt.figure(2)
        plt.clf()
        
        # Plot robots
        for i in range(robot_N):
            if Eff_robot[i] == 1:
                plt.plot(position[i, 0], position[i, 1], 'go')
                plt.plot(150, 250, 'ro')  # goal position
                
                # Plot orientation indicator
                Dir_r = 4
                line_x = [position[i, 0], position[i, 0] + Dir_r * np.cos(position[i, 2])]
                line_y = [position[i, 1], position[i, 1] + Dir_r * np.sin(position[i, 2])]
                plt.plot(line_x, line_y, color=[0, 1, 0])
                plt.text(position[i, 0] + 5, position[i, 1], str(i))
            else:
                plt.plot(position[i, 0], position[i, 1], 'rx')
                plt.text(position[i, 0] + 5, position[i, 1], str(i))
        
        plt.title(f'Robot positions (Mode {MODE})')
        plt.draw()
        plt.pause(0.001)
    
    return dis

# Helper functions that would need to be implemented
def Map_creator(N):
    """Creates a map with obstacles"""
    # This should be implemented based on your specific map requirements
    return np.zeros((N, N))

def PlotLine(robot, sensor, map, MODE):
    """Simulates laser scanning and updates map"""
    # This should be implemented based on your sensor model
    dis = np.array([100.0] * 5)  # Dummy distances
    return map, dis

def PlotMap(map):
    """Calculates coverage metrics"""
    # This should be implemented based on your coverage calculation
    return 0, 0, 0, 0  # Dummy return values